### PR TITLE
DKI and mean signal DKI script

### PR DESCRIPTION
**Pull request by mdesco**
DKI computation from DIPY. This is heavily based on two DIPY reconst examples, wrapped in the usual SCILPY way. DKI is more vulnerable to noise than I was expecting but this is a good HARDI addition to our scripts.

Data to test this PR can we download here:
https://www.dropbox.com/s/95b7qwi09ok28oi/DKI_PR.zip?dl=0

cd DKI_PR
./run.sh

### CODE REVIEW

File: scripts/scil_compute_kurtosis_metrics.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+from scilpy.io.utils import (add_overwrite_arg, assert_inputs_exist,
+                             assert_outputs_exist, add_force_b0_arg)
+from scilpy.utils.bvec_bval_tools import (normalize_bvecs, is_normalized_bvecs,
+                                          check_b0_threshold)
+
+logger = logging.getLogger("Compute_DKI_Metrics")
+logger.setLevel(logging.INFO)
+
+
+def _guess_bvals_centroids(bvals, threshold):
```
Comment (GuillaumeTh) [inline scripts/scil_compute_kurtosis_metrics.py]: Can you add this function in scilpy/utils/bvec_bval_tools.py and can you add the documentation to the function please

File: scripts/scil_compute_kurtosis_metrics.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+    for bval in bvals[1:]:
+        diffs = np.abs(np.asarray(bval_centroids) - bval)
+        # Found no bval in bval centroids close enough to the current one.
+        if not len(np.where(diffs < threshold)[0]):
+            bval_centroids.append(bval)
+
+    return np.array(bval_centroids)
+
+
+def identify_shells(bvals, threshold=40.0):
```
Comment (GuillaumeTh) [inline scripts/scil_compute_kurtosis_metrics.py]: Can you add this function in scilpy/utils/bvec_bval_tools.py and can you add the documentation to the function please

File: scripts/scil_compute_kurtosis_metrics.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+
+from scipy.ndimage.filters import gaussian_filter
+
+# Aliased to avoid clashes with images called mode.
+from scilpy.io.utils import (add_overwrite_arg, assert_inputs_exist,
+                             assert_outputs_exist, add_force_b0_arg)
+from scilpy.utils.bvec_bval_tools import (normalize_bvecs, is_normalized_bvecs,
+                                          check_b0_threshold)
+
+logger = logging.getLogger("Compute_DKI_Metrics")
```
Comment (GuillaumeTh) [inline scripts/scil_compute_kurtosis_metrics.py]: Please add these two lines in the main function

File: scripts/scil_compute_kurtosis_metrics.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+
+import dipy.reconst.dki as dki
+import dipy.reconst.msdki as msdki
+
+from dipy.io.gradients import read_bvals_bvecs
+from dipy.core.gradients import gradient_table
+
+from scipy.ndimage.filters import gaussian_filter
+
+# Aliased to avoid clashes with images called mode.
```
Comment (GuillaumeTh) [inline scripts/scil_compute_kurtosis_metrics.py]: Can you remove the comment please ?

File: scripts/scil_compute_kurtosis_metrics.py [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+                   help='Path to a binary mask.' +
+                   '\nOnly data inside the mask will be used '
+                   'for computations and reconstruction. ' +
+                   '\n(Default: None)')
+
+    p.add_argument('--tolerance', '-t',
+                   metavar='INT', type=int, default=20,
+                   help='The tolerated distance between the b-values to '
+                   'extract\nand the actual b-values [Default: %(default)s].')
+    p.add_argument('--min_k', dest='min_k', type=float, default='0',
```
Comment (GuillaumeTh) [inline scripts/scil_compute_kurtosis_metrics.py]: `default='0'` -> `default=0.0`

File: scripts/scil_compute_kurtosis_metrics.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+                   help='Minium kurtosis value in the output maps ' +
+                   '\n(ak, mk, rk). In theory, -3/7 is the min kurtosis ' +
+                   '\n limit for regions that consist of water confined ' +
+                   '\nto spherical pores (see DIPY example and ' +
+                   '\n documentation) [Default: %(default)s].')
+    p.add_argument('--max_k', dest='max_k', type=float, default='3',
```
Comment (GuillaumeTh) [inline scripts/scil_compute_kurtosis_metrics.py]: `default='3'` -> `default=3.0`

File: scripts/scil_compute_kurtosis_metrics.py [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+                   help='Maximum kurtosis value in the output maps ' +
+                   '\n(ak, mk, rk). In theory, 10 is the max kurtosis' +
+                   '\nlimit for regions that consist of water confined' +
+                   '\nto spherical pores (see DIPY example and ' +
+                   '\ndocumentation) [Default: %(default)s].')
+    p.add_argument('--smooth', dest='smooth', type=float, default='2.5',
```
Comment (GuillaumeTh) [inline scripts/scil_compute_kurtosis_metrics.py]: `default='2.5'` -> `default=2.5`

File: scripts/scil_compute_kurtosis_metrics.py (hunk 6 above)
Comment (GuillaumeTh) [reply scripts/scil_compute_kurtosis_metrics.py]: Remove `dest` option

File: scripts/scil_compute_kurtosis_metrics.py (hunk 5 above)
Comment (GuillaumeTh) [reply scripts/scil_compute_kurtosis_metrics.py]: Remove `dest` option. The default destination is `min_k`

File: scripts/scil_compute_kurtosis_metrics.py (hunk 7 above)
Comment (GuillaumeTh) [reply scripts/scil_compute_kurtosis_metrics.py]: Remove `dest` option

File: scripts/scil_compute_kurtosis_metrics.py [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+                   '\nsensitive and outliers occur easily. According to' +
+                   '\ntests on HCP, CB_Brain, Penthera3T, this smoothing' +
+                   '\nis thus turned ON by default with fwhm=2.5. ' +
+                   '\n[Default: %(default)s].')
+    p.add_argument('--not_all', action='store_true', dest='not_all',
+                   help='If set, will only save the metrics explicitly ' +
+                   '\nspecified using the other metrics flags. ' +
+                   '\n[Default: not set].')
+
+    g = p.add_argument_group(title='Metrics files flags')
```
Comment (GuillaumeTh) [inline scripts/scil_compute_kurtosis_metrics.py]: For the next options, remove dest option if `dest` is the same name than the option

**Comment by mdesco** [comment]
I will have a hard time to work on these suggestions quickly... Sorry...

File: scripts/scil_compute_kurtosis_metrics.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+                   '\nOnly data inside the mask will be used '
+                   'for computations and reconstruction. ' +
+                   '\n(Default: None)')
+
+    p.add_argument('--tolerance', '-t',
+                   metavar='INT', type=int, default=20,
+                   help='The tolerated distance between the b-values to '
+                   'extract\nand the actual b-values [Default: %(default)s].')
+    p.add_argument('--min_k', type=float, default=0.0,
+                   help='Minium kurtosis value in the output maps ' +
```
Comment (karanphil) [inline scripts/scil_compute_kurtosis_metrics.py]: Correct typo: Minium -> Minimum

Comment (karanphil) [reply scripts/scil_compute_kurtosis_metrics.py]: Also remove space before "limit" and "documention" just below.

File: scripts/scil_compute_kurtosis_metrics.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+    # Validate bvals and bvecs
+    bvals, bvecs = read_bvals_bvecs(args.bvals, args.bvecs)
+    if not is_normalized_bvecs(bvecs):
+        logging.warning('Your b-vectors do not seem normalized...')
+        bvecs = normalize_bvecs(bvecs)
+
+    # Find the volume indices that correspond to the shells to extract.
+    tol = args.tolerance
+    shells, _ = identify_shells(bvals, tol)
+    print(shells)
```
Comment (GuillaumeTh) [inline scripts/scil_compute_kurtosis_metrics.py]: Please remove this print.

**Comment by jchoude** [comment]
I made a quick single PR on this branch @mdesco . Once merged (on your page), we can merge this.

**Pull request merged**